Change coord transform from 1/sigma^2 to 1/sigma in Gaussian

RichardWaiteSTFC opened the pull request:
### Description of work

<!--  OLD DESCRIPTION
There is a bug somewhere - I suspect that the coordinate transform used in jacobian is not propagated through the LM correctly when the parameter is updated. Perhaps someone is calling `getParameter` when they should call `activeParameter` or similar?~

Our ideal solution would be to get it working with the coordinate transform used. There is a note in the docs which implies it should be more stable.
https://github.com/mantidproject/mantid/blob/28108db45d4287506f72cae1b3ddaaddd0a2784f/docs/source/tutorials/extending_mantid_with_python/advanced_fit_behaviours/01_fit_stability.rst?plain=1#L20-L22

Another solution would be to make a second Gaussian function e.g. `GaussianInverseVariance` or similar that has 1/sig^2 as a parameter (no internal coord transforms).~
-->

See comments below for a blow by blow...
This PR does the following:
1.  Change the coordinate transform in the gaussian function from `1/sigma^2` to `1/sigma`.
2.  Also normalises `histogram1D` and `histogramDerivative1D` by bin width (this seems to be a bug with other functions as well e.g. `FlatBackground`) which I will fix separately so i don't get confused about which change is causing which test to fail. 


### To test:
(1) Run this script
```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from itertools import cycle

ws = CreateSampleWorkspace(OutputWorkspace='ws', Function='User Defined', UserDefinedFunction='name=Gaussian,Height=100,PeakCentre=5,Sigma=0.6;name=FlatBackground,A0=10', 
                           XMin=1, XMax=10, BinWidth=0.1, NumBanks=1, BankPixelWidth=1, PixelSpacing=0.02)

initial_sigma = 0.2  # try also 0.7

num_derivs = ["true", "false"]
eval_types = ["CentrePoint", "Histogram"]
colors = iter(["tab:blue", "tab:orange", "tab:red", "tab:green"])
lstyles = cycle(['-', '--', '-.', ':'])

fig, axes = plt.subplots(dpi=156.0, edgecolor='#ffffff', num='ws_Workspace-1', subplot_kw={'projection': 'mantid'})
axes.errorbar(ws, color='k', marker='o', markersize=3, capsize=2, ls='', alpha=0.5, label='data', wkspIndex=0)
heights = []
height_errors = []
for eval_type in eval_types:
    for num_deriv in num_derivs:
        # fit
        func_str = f"composite=CompositeFunction,NumDeriv={num_deriv};name=Gaussian,Height=60,PeakCentre=5,Sigma={initial_sigma};name=FlatBackground,A0=10"
        out_wsname = f"NumDeriv={num_deriv}, Eval={eval_type}"
        res = Fit(Function=func_str, 
                  InputWorkspace=ws, MaxIterations=20000, 
                  Minimizer='Levenberg-Marquardt', EvaluationType=eval_type,
                  Output=out_wsname, OutputCompositeMembers=True, StartX=1.0945, EndX=9.9055)
        # save heights
        heights.append(res.Function["f0.Height"])
        height_errors.append(res.Function.getError("f0.Height"))
        # plot fit
        axes.plot(res.OutputWorkspace, color=next(colors), ls=next(lstyles), alpha=0.7, wkspIndex=1, label = out_wsname)
    # plot initial guess (depends on eval type)
    ws_eval = EvaluateFunction(Function=func_str, InputWorkspace=ws, OutputWorkspace=out_wsname + "_eval", EvaluationType=eval_type)
    axes.plot(ws_eval, color='k', ls=next(lstyles), label = out_wsname + " (Initial)", wkspIndex=1)  # fit curve
legend = axes.legend(fontsize=8.0).set_draggable(True).legend
fig.show()


print("NumDeriv\tEval\tHeight\tError")
ii = 0
for eval_type in eval_types:
    for num_deriv in num_derivs:
        print(f"{num_deriv}\t{eval_type}\t{heights[ii]:.2f}\t{height_errors[ii]:.2f}")
        ii += 1
```

<img width="967" height="781" alt="image" src="https://github.com/user-attachments/assets/5b5378da-f957-485c-8c5f-4b9ce5eca4d6" />

Note that all fits converge on expected solution (due to fix 1)

Also note that the peak height of the initial guess is the same in both `EvaluationType="Histogram"` and `"CentrePoint"` due to fix (2).

I think there is a similar problem (missing normalisation by bin width in `FlatBackground` (and possibly other functions) which means the initial guess for  `EvaluationType="Histogram"` has a background that is factor ~10 too small (bin width is 0.1). Not sure why you would need histogram evaluation for a flat background as it should be equivalent to the centre point...

The script print out this on main
```
NumDeriv	Eval	Height	Error
true	CentrePoint	81.79	5.78
false	CentrePoint	81.79	5.78
true	Histogram	4498.23	2114150516.50
false	Histogram	4496.88	2061130372.28
```

and on this branch (note errors agree with scipy - see https://github.com/mantidproject/mantid/pull/40267#issuecomment-3490487656)
```
NumDeriv	Eval	Height	Error
true	CentrePoint	100.00	3.52
false	CentrePoint	100.00	3.52
true	Histogram	100.12	3.52
false	Histogram	100.12	3.52
```

2. Check that histogram mode is working as expected by running this script
```
bin_widths = np.arange(0.1,1.6,0.1)
sigma = 0.6
ws = CreateSampleWorkspace(OutputWorkspace='ws', Function='User Defined', 
                           UserDefinedFunction=f'name=Gaussian,Height=100,PeakCentre=5,Sigma={sigma};name=FlatBackground,A0=10', 
                           XMin=1, XMax=10, BinWidth=0.1, NumBanks=1, BankPixelWidth=1, PixelSpacing=0.02)


num_derivs = ["true", "false"]
eval_types = ["CentrePoint", "Histogram"]
par_names = ["f0.Height", "f0.PeakCentre", "f0.Sigma", "f1.A0"]
pfits = {par: np.zeros((len(num_derivs), len(eval_types), len(bin_widths))) for par in par_names}

for iwidth, bin_width in enumerate(bin_widths):
    ws_rb = Rebin(InputWorkspace=ws, Params=bin_width)
    ws_rb /= (bin_width/0.1)
    for ieval, eval_type in enumerate(eval_types):
        for ideriv, num_deriv in enumerate(num_derivs):
            # fit
            func_str = f"composite=CompositeFunction,NumDeriv={num_deriv};name=Gaussian,Height=60,PeakCentre=5,Sigma=0.7;name=FlatBackground,A0=10"
            out_wsname = f"NumDeriv={num_deriv}_Eval={eval_type}_width{bin_width:.1f}"
            res = Fit(Function=func_str, 
                      InputWorkspace=ws_rb, MaxIterations=20000, 
                      Minimizer='Levenberg-Marquardt', EvaluationType=eval_type,Normalize=False,
                      Output=out_wsname, OutputCompositeMembers=False, StartX=1.0945, EndX=9.9055)
            # save heights
            for par in par_names:
                pfits[par][ideriv, ieval, iwidth] = abs(res.Function[par])
     
fig, ax = plt.subplots(2,1)
for iax, ipar in enumerate([0, 2]):
    colors = iter(["tab:blue", "tab:orange", "tab:red", "tab:green"])
    lstyles = cycle(['-', '--', '-.', ':'])
    par_name = par_names[ipar]
    for ieval, eval_type in enumerate(eval_types):
        for ideriv, num_deriv in enumerate(num_derivs):
            label = f"NumDeriv={num_deriv}, Eval={eval_type}"
            ax[iax].plot(bin_widths/sigma, pfits[par_name][ideriv, ieval, :], color=next(colors), ls=next(lstyles), alpha=0.7,  label=label)
    legend = ax[iax].legend(fontsize=8.0).set_draggable(True).legend
    ax[iax].set_xlabel("Bin Width/Sigma")
    ax[iax].set_ylabel(par_name)
fig.show()
```
<img width="1348" height="511" alt="image" src="https://github.com/user-attachments/assets/c50e9ecd-3d48-44a0-84d4-398af390fc93" />

You can see histogram mode fit parameter  are stable and agree with input (height=100, sigma=0.6) - as bin width increases relative to sigma the centre-point evaluation overestimates sigma (as expected).


Closes #40293

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?
- If the PR author isn’t in the `mantid-developers` or `mantid-contributors` teams, add a review comment `rerun ci` to authorize/rerun the CI

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

RichardWaiteSTFC committed:
Remove coord transform from sigma to 1/sigma^2

RichardWaiteSTFC committed:
Use 1/sigma transform not 1/sigma^2

Think problems previously if 1/sigma^2 went negative

RichardWaiteSTFC committed:
Update histogram jacobian calculation

RichardWaiteSTFC committed:
Normalise histogram 1D and jacobian by bin width

Looks like this is a bug in FlatBackground as well

RichardWaiteSTFC committed:
Update functionLocal to use weight = 1/sigma

Not strictly necessary, but for consistency

RichardWaiteSTFC committed:
Fix bug in histogram evaluation of jacobian

RichardWaiteSTFC committed:
Fix test values for height and sigma in FindEPP

Centre unchanged and chi-squared of fit same as on main to within 6 decimal points (note the unit test asserts chi-squared to 4 decimal points - and that line did not cause a failure).
https://github.com/mantidproject/mantid/blob/c91f05d6ee4a4c60fc552671bae075047ff1cb7e/Framework/Algorithms/test/FindEPPTest.h#L145

New parameters are negligibly closer to actual peak parameter used to simulate data, but the fit is unrealistic anyway as it doesn't account for the FlatBackground present in the data but not in model.

RichardWaiteSTFC committed:
Fix height in FitPeakTest FitPeakOneStep & FitPeakOneStepFullPeakName

Fit clearly better on this branch - data generated by gen_PG3DiamondData
exhibits a peak of rough height ~40. The fit now gets height of ~38 rather than ~26.

Indeed, the fit is now much more similar to starting parameters
https://github.com/mantidproject/mantid/blob/c91f05d6ee4a4c60fc552671bae075047ff1cb7e/Framework/Algorithms/test/FitPeakTest.h#L236

RichardWaiteSTFC committed:
Update intens values in HB3AIntegrateDetectorPeaksTest

Small changes (typically ~4 dp) in values ~1000.
Visual inspection of fits seem equally good

RichardWaiteSTFC committed:
Update corelli calibration test_bank

Improvements to fitting Gaussian peaks seem to have produced fewer 'outlier' tubes that require masking.
The cost function of the fits is better in all cases. In the tubes that were previously masked but now not identified as outliers, the deviation in the vertical positions from the average for all tubes in the bank (see function criterion_peak_vertical_position) was consistent with other valid tubes.

RichardWaiteSTFC committed:
Fix ht in FitPeakTest FitPeakOneStepFullPeakName (missed last time)

RichardWaiteSTFC commented:
Progress so far:

- Fixed FitPeaksTest tests - fit now much better
<img width="455" height="165" alt="image" src="https://github.com/user-attachments/assets/0f2014a5-3d5a-4a22-898d-71127548de10" />

- Fixed FindEPPTest (chi-squared of fit same as on main to within 6 decimal points)

- Fixed HB3AIntegrateDetectorPeaksTest - small changes to intensities, visual inspected fit and looks good
<img width="1137" height="307" alt="image" src="https://github.com/user-attachments/assets/ab251f3c-3df0-4d21-acaf-3edc261dd9c2" />

- Fixed corelli calibration test_bank - cost function of fits seems to be improved and fewer 'outlier' tubes that require masking. BUt it would be great is someone (@zjmorgan?) could check that the quality of the calibration is indeed better (or at least not worse!) and the subsequent reduction is not affected.

zjmorgan commented:
> Progress so far:
> 
> * Fixed FitPeaksTest tests - fit now much better
> 
> <img alt="image" width="455" height="165" src="https://private-user-images.githubusercontent.com/55979119/514538255-0f2014a5-3d5a-4a22-898d-71127548de10.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMxNjM5NDAsIm5iZiI6MTc2MzE2MzY0MCwicGF0aCI6Ii81NTk3OTExOS81MTQ1MzgyNTUtMGYyMDE0YTUtM2Q1YS00YTIyLTg5OGQtNzExMjc1NDhkZTEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTE0VDIzNDA0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiNWI0M2Q0MGI3ODcxNDE2YTg2ZDc4NDA1MDNkMWY4NDNlYmIwNGNhMDE2YmE1MDY5YTlkMTFmN2ZkYzFmMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ODRki6a27Vnez9AmfrgcCBQldRWSjXqrRBz1dUDvN7E">
> * Fixed FindEPPTest (chi-squared of fit same as on main to within 6 decimal points)
> * Fixed HB3AIntegrateDetectorPeaksTest - small changes to intensities, visual inspected fit and looks good
> 
> <img alt="image" width="1137" height="307" src="https://private-user-images.githubusercontent.com/55979119/514539035-ab251f3c-3df0-4d21-acaf-3edc261dd9c2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMxNjM5NDAsIm5iZiI6MTc2MzE2MzY0MCwicGF0aCI6Ii81NTk3OTExOS81MTQ1MzkwMzUtYWIyNTFmM2MtM2RmMC00ZDIxLWFjYWYtM2VkYzI2MWRkOWMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTE0VDIzNDA0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZDBlN2I4MmJlNjRmMzBjM2YwYzcwMmJjNWMzZDFmYWJiNThiN2FlNWFkNjIwNzI1MWZiYWY4YjBhMDE3YzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Q-Rmp59W81ggl-QyhS-JlwUiQvxMBfvtMkWfgZY49cQ">
> * Fixed corelli calibration test_bank - cost function of fits seems to be improved and fewer 'outlier' tubes that require masking. BUt it would be great is someone (@zjmorgan?) could check that the quality of the calibration is indeed better (or at least not worse!) and the subsequent reduction is not affected.

This looks good to me!

RichardWaiteSTFC committed:
Fix bug in jacobian of weight for histogram

RichardWaiteSTFC committed:
Fix FitFunctionTest and add test Histo and CentrePoint jac agreement

RichardWaiteSTFC committed:
Update offsets in GetDetectorOffsetsTest

Offsets now agree with equations in the docs for a peak centred at 1 (which is the case for the simulated data).

RichardWaiteSTFC committed:
Update peak centre in RectanglePLotTest

Peak centre change 1e-4 smaller than peak width and peak x and y position are still the same as expected.

RichardWaiteSTFC committed:
Normalise histo values by bin width in HistogramDomainCreatorTest

A better solution would be to do this normalisation int he function that simulates the data
https://github.com/mantidproject/mantid/blob/b47cbe1dc3c8af5cda1f80bb9adf3b716ba4e2cc/Framework/CurveFitting/test/HistogramDomainCreatorTest.h#L454
But unfortunately other functions with histo evaluation (e.g. Lorentzian that have not had the histogram 1D function updated as in Gaussian).

RichardWaiteSTFC committed:
Remove unused variable

RichardWaiteSTFC committed:
Add release note and update docs

RichardWaiteSTFC committed:
Fix clang error unused variable

RichardWaiteSTFC committed:
Fix failing GetDetectorOffsetsTest on Linux

Issue is that it was no longer reliably failing to fit!

RichardWaiteSTFC committed:
Tidy up/clarify GetDetectorOffsetTest

Combine testExecMasksAreConsistentWithDetectorFlags and testExecMasksAreCombined which differed only by an assert

RichardWaiteSTFC committed:
Update fit parameters in WeightedLeastSquaresTest

Fit is similar to main - slightly improved chi-squared on this branch (0.97) vs. main (1.03)

RichardWaiteSTFC committed:
Update a in POLDIFitPeaks2DTest and POLDIDataAnalysisTest

Mean of error weighed residuals is slightly improved on this branch (2.4611) vs main (2.4613).
In addition fit converged more quickly on main (half execution time).

RichardWaiteSTFC commented:
@perenon - I'm having some trouble with an ILL system test `ILLDirectGeometryReductionTest.IN4 `
We get an error from `DirectILLIntegrateVanadium > ComputeCalibrationCoefVan > Integration`  where it tries to integrate +/- N*sigma around elastic line but we end up getting Max<Min as the fit returns a negative sigma by chance on this branch

<img width="554" height="126" alt="image" src="https://github.com/user-attachments/assets/83fd65f6-15de-4e45-aa74-45be7f412b3f" />

For now my plan is to take the absolute value of sigma here
https://github.com/mantidproject/mantid/blob/5f0144cb858a0c2b495836d9d674c8387700702e/Framework/Algorithms/src/IntegrateEPP.cpp#L90
I could also constrain `Sigma>0` here
https://github.com/mantidproject/mantid/blob/52f4dc96bd1c5828c1014671243f21dd86b0749e/Framework/Algorithms/src/FindEPP.cpp#L124-L126
Do you have a preference?

Perhaps the real problem is that the reduction tries to use sigma from a Gaussian fit to this data
<img width="550" height="360" alt="image" src="https://github.com/user-attachments/assets/30ce70b6-e0d8-4af5-9ab8-14311365c013" />

I think long term `FindEPP` needs some work. Ideally it should only attempt to fit if there are sufficient counts, but this can be difficult to quantify. If you expect elastic line to be at roughly same TOF for all spectra you could fit sum over all detectors to get  an initial integration range, call `Integration` and only fit spectra with intensity/error > N? As you have the raw counts here you could say you need N bins with counts > 10 (but don't know if raw counts are always passed to `FindEPP`). Or some post fitting success criteria can be defined (either in `FindEPP` or `DirectILLCollectData` if you only want to touch ILL stuff). For example exclude fits by ratio of parameter to error for peak height and sigma? Also put limits on chi-squared? But I think that goes beyond this PR if you are OK with that?

RichardWaiteSTFC committed:
Update tolerance on ILLIndirectReductionQENSTest

Output looks comparable - but checking with ILL

RichardWaiteSTFC committed:
Take abs(sigma) to stop Max < Min range passed to Integration

Relates to ILLDirectGeometryReductionTest.IN4
We get an error from DirectILLIntegrateVanadium > ComputeCalibrationCoefVan > Integration where it tries to integrate +/- N*sigma around elastic line but we end up getting Max<Min as the fit returns a negative sigma by chance on this branch

RichardWaiteSTFC committed:
Change name of test so mismatch file not overwritten

Mismatch filename of equivalent failing test for GEM has same name.
Can revert.

RichardWaiteSTFC committed:
Set sigma to be positive in FindEPP

RichardWaiteSTFC committed:
Update ref. files for focussed data in PEARL and GEM CreateCalTest

RichardWaiteSTFC committed:
Update reference file for ILLDirectGeometryReductionTest.IN4

Differences of order 1e-5 over intensity of order ~ 250.

jhaigh0 reviewed (approved):
Code and resulting improvements are both looking good to me 👍
